### PR TITLE
Changed constructor to use sparsity_pattern() for sparse matrices

### PR DESCRIPTION
**Pull request by SamuelzXu**
Attempt at closing #871 with help from @YingboMa

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by SamuelzXu** [comment]
@ChrisRackauckas how do I approach the SafeTestSets bug?

**Comment by ChrisRackauckas** [comment]
What Safetestsets bug? I don't see one.

The error seems to be that you need to change the matrix from a matrix of booleans to a matrix of numbers.

File: src/systems/diffeqs/abstractodesystem.jl [hunk 1]
```diff
@@ -220,7 +220,7 @@ function DiffEqBase.ODEFunction{iip}(sys::AbstractODESystem, dvs = states(sys),
                      jac = _jac === nothing ? nothing : _jac,
                      tgrad = _tgrad === nothing ? nothing : _tgrad,
                      mass_matrix = _M,
-                     jac_prototype = sparse ? similar(get_jac(sys)[],Float64) : nothing,
+                     jac_prototype = sparse ? jacobian_sparsity(sys) : nothing, # similar(sys.jac[],Float64)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:223]:
```suggestion
                     jac_prototype = sparse ? convert.((eltype(u0),),jacobian_sparsity(sys)) : nothing, # similar(sys.jac[],Float64)
```

Comment (ChrisRackauckas) [reply src/systems/diffeqs/abstractodesystem.jl:223]: Actually hmm, how do you do that without a `u0` type? I guess you just guess Float64 for now?

**Review by YingboMa — CHANGES_REQUESTED** [review]
This is incorrect. We want use `sparsity_pattern` only when `jac=false` and `sparse=true`.

File: src/systems/diffeqs/abstractodesystem.jl [hunk 2]
```diff
@@ -26,7 +26,7 @@ function calculate_jacobian(sys::AbstractODESystem;
     dvs = states(sys)
 
     if sparse
-        jac = sparsejacobian(rhs, dvs, simplify=simplify)
+        jac = sparsity_pattern(sys)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:29]: This should probably be reverted.

**Comment by ChrisRackauckas** [comment]
I think you can achieve that by just using `jacobian_sparsity(sys)` in the ODEFunction construction? The issue from before was that `get_jac(sys)` would be building the full version just to create the Jacobian, which ended up full.

**Comment by YingboMa** [comment]
You should also promote the `jac_prototype` to have the same `eltype` with `u0` in `ODEProblem` when you have `u0`.

**Comment by SamuelzXu** [comment]
@YingboMa Hopefully this does the trick

**Pull request closed without merging**